core/imds/src/main/java/software/amazon/awssdk/imds/internal/DefaultEc2MetadataAsyncClient.java

```diff
@@ -0,0 +1,307 @@
+        this.httpClient = builder.httpClient != null ? builder.httpClient
+                                                     : AwsCrtAsyncHttpClient.create();
+        this.endpointMode = builder.endpointMode != null ? builder.endpointMode
+                                                         : DEFAULT_ENDPOINT_PROVIDER.resolveEndpointMode();
+        this.endpoint = builder.endpoint != null ? builder.endpoint
+                                                 : URI.create(DEFAULT_ENDPOINT_PROVIDER.resolveEndpoint(this.endpointMode));
+        this.requestMarshaller = new RequestMarshaller(this.endpoint);
+        this.tokenTtl = builder.tokenTtl != null ? builder.tokenTtl
+                                                 : Duration.ofSeconds(21600);
+        this.asyncRetryScheduler = builder.scheduledExecutorService != null
+                                   ? builder.scheduledExecutorService
+                                   : Executors.newScheduledThreadPool(DEFAULT_RETRY_THREAD_POOL_SIZE);
+    }
+
+    public static Ec2MetadataAsyncClient.Builder builder() {
+        return new DefaultEc2MetadataAsyncClient.Ec2MetadataAsyncBuilder();
+    }
+
+    @Override
+    public CompletableFuture<MetadataResponse> get(String path) {
+        return get(path, RetryPolicyContext.builder().retriesAttempted(0).build());
+    }
+
+    private CompletableFuture<MetadataResponse> get(String path, RetryPolicyContext retryPolicyContext) {
+        SdkHttpFullRequest baseTokenRequest = requestMarshaller.createTokenRequest(tokenTtl);
+
+        CompletableFuture<MetadataResponse> result = createAsyncRequest(baseTokenRequest)
+            .thenCompose(token -> {
+                SdkHttpFullRequest baseMetadataRequest = requestMarshaller.createDataRequest(path, token, tokenTtl);
+                return createAsyncRequest(baseMetadataRequest);
+            }).thenApply(MetadataResponse::create);
+
+        return result.handle((response, error) -> {
+            if (response != null || !shouldRetry(retryPolicyContext, error)) {
+                return result;
+            }
+            int newAttempt = retryPolicyContext.retriesAttempted() + 1;
+            log.debug(() -> "Retrying request: Attempt " + newAttempt);
+            RetryPolicyContext newContext =
+                RetryPolicyContext.builder()
+                                  .request(baseTokenRequest)
+                                  .retriesAttempted(newAttempt)
+                                  .exception(SdkClientException.create(error.getMessage(), error))
+                                  .build();
+            return scheduledRetryAttempt(path, newContext);
+        }).thenCompose(Function.identity()); // only java 12 has .exceptionallyCompose()
+    }
+
+    private CompletableFuture<String> createAsyncRequest(SdkHttpFullRequest baseRequest) {
+        StringAsyncResponseHandler responseHandler = new StringAsyncResponseHandler();
```

Comment:
do we need to create a new instance of the response handler for each request attempt?

Reply:
Yes we do, because the handler contains the future that has the actual value of the request.